### ISSUE TITLE
[java] UnusedAssignment false positives

### ISSUE THREAD

**Issue by davidburstrom** [pmd/pmd#2668]
**Affects PMD Version:** 

6.26.0

**Description:**

The [UnusedAssignment](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#unusedassignment) rule flags unused assignments for fields, but the fields might be used in the program flow when invoking methods in between assignments.

**Code Sample demonstrating the issue:**

Suppose the `Worker` class calls the listener in its constructor (it is of course questionable to begin with, but something similar is used in a real-life scenario):

```java
class A implements Worker.Listener {
  private boolean ignore;
  private Worker worker;

  A() {
    ignore = true;
    worker = new Worker(this);
    ignore = false;
  }

  void doWork() {
    worker.work();
  }

  public void onWork() {
    if (ignore) {
      return;
    }
    System.out.println("onWork");
  }
}
```

This reports

```console
A.java:6:	The value assigned to field 'ignore' is never used (overwritten on line 8)
```

**Steps to reproduce:**

See code sample.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

N/A

**Comment by oowekyala** [comment]
Thanks for the report. It looks like we need to check for `this` leaks in constructors.
